silhouette/app/com/mohiva/play/silhouette/impl/providers/oauth2/Auth0Provider.scala

```diff
@@ -0,0 +1,198 @@
+package com.mohiva.play.silhouette.impl.providers.oauth2
+
+import com.mohiva.play.silhouette.api.LoginInfo
+import com.mohiva.play.silhouette.api.util.HTTPLayer
+import com.mohiva.play.silhouette.impl.exceptions.ProfileRetrievalException
+import com.mohiva.play.silhouette.impl.providers._
+import com.mohiva.play.silhouette.impl.providers.oauth2.Auth0Provider._
+import play.api.libs.json.{ JsString, JsValue }
+import play.api.mvc.RequestHeader
+import play.mvc.Http
+
+import scala.concurrent.Future
+
+/**
+ * Base Auth0 OAuth2 Provider.
+ *
+ * OAuth Provider configuration in silhouette.conf must indicate:
+ *
+ *   # Auth0 Service URLs
+ *   auth0.authorizationURL="https://mydomain.eu.auth0.com/authorize"
+ *   auth0.accessTokenURL="https://mydomain.eu.auth0.com/oauth/token"
+ *   auth0.apiURL="https://mydomain.eu.auth0.com/userinfo"
+ *
+ *   # Application URL and credentials
+ *   auth0.redirectURL="http://localhost:9000/authenticate/auth0"
+ *   auth0.clientID=myoauthclientid
+ *   auth0.clientSecret=myoauthclientsecret
+ *
+ *   # Auth0 user's profile information requested
+ *   auth0.scope="openid name email picture"
+ *
+ * See http://auth0.com for more information on the Auth0 Auth 2.0 Provider and Service.
+ */
+trait BaseAuth0Provider extends OAuth2Provider {
+
+  /**
+   * The content type to parse a profile from.
+   */
+  override type Content = JsValue
+
+  /**
+   * The provider ID.
+   */
+  override val id = ID
+
+  /**
+   * Defines the URLs that are needed to retrieve the profile data.
+   */
+  override protected val urls = Map("api" -> settings.apiURL.getOrElse(API))
+
+  /**
+   * Builds the social profile.
+   *
+   * @param authInfo The auth info received from the provider.
+   * @return On success the build social profile, otherwise a failure.
+   */
+  override protected def buildProfile(authInfo: OAuth2Info): Future[Profile] = {
+    val request = httpLayer.url(urls("api"))
+    val requestWithHeader = request.withHeaders(("Authorization", s"Bearer ${authInfo.accessToken}"))
+
+    val httpResponse = requestWithHeader.get()
+    httpResponse.flatMap { response =>
+      response.status match {
+        case Http.Status.OK =>
+          profileParser.parse(response.json, authInfo)
+
+        case _ => {
+          val json = response.json
+          (json \ "error").asOpt[JsString] match {
+            case Some(error) =>
+              val errorDescription = (json \ "error_description")
+              throw new ProfileRetrievalException(SpecifiedProfileError.format(id, errorDescription, error, response.status))
+            case _ =>
+              throw new ProfileRetrievalException(s"Cannot get user profile - HTTP status code ${response.status}")
```

Comment:
Could you please use a constant for the error message. So the exception case can be better tested.